File: js/019_RemoveNthNodeFromEndofList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    var pre = new ListNode(-1);
    pre.next = head;
    var p = pre;
    var q = head;
    for(var i = 0; i < n-1; i++){
        q = q.next;
    }
    while(q.next !== null){
        p = p.next;
        q = q.next;
    }
    p.next = p.next.next;
    return pre.next;
};

File: cpp/019_RemoveNthNodeFromEndofList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode **p = &head, *q = head;
        for(int i = 0; i < n-1; ++i){
            q = q->next;
        }
        while( q->next != NULL ){
            p = &((*p)->next);
            q = q->next;
        }
        *p = (*p)->next;
        return head;
    }
};